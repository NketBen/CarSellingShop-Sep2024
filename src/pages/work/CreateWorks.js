
import { useState, useEffect } from "react";
import { db } from "../../firebase";
import { useNavigate, useParams } from "react-router-dom";
import { collection, addDoc, serverTimestamp } from "firebase/firestore";
import {
  getStorage,
  ref,
  uploadBytesResumable,
  getDownloadURL,
} from "firebase/storage";
import "bootstrap/dist/css/bootstrap.min.css";
import Nav from "react-bootstrap/Nav";
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";
import InputGroup from "react-bootstrap/InputGroup";
import { getAuth} from "firebase/auth";
import { v4 as uuidv4 } from "uuid";
import { toast } from "react-toastify";

export default function CreateWorks() {
  // const [newName, setNewName] = useState("");
  // const [newDepart, setNewDepart] = useState("");
  // const [newTarea, setNewTarea] = useState("");
  // const [newForm, setNewForm] = useState("");
  // const [estado, setEstado] = useState("");
  // const [urgencia, setUrgencia] = useState("");
  // const [creator, setCreator] =useState("");
   const workCollectionRef = collection(db, "work");
   const [startDate, setStartDate] = useState(new Date());
  const [endDate, setEndDate] = useState(new Date());
  const [loading, setLoading] = useState(false);
  //const[images, setImages]=useState([]);

const [paraForm, SetParaForm] = useState({
    
    contracta: false,
    address: "no contratada",
    telNo:+34,
    contractaName:"no contratada",
    newName:"",
newDepart:"",
newTarea:"",
newForm:"",
estado:"",
urgencia:"",
creator:"",

images:{},
  });

  const {
   // images,
    address,
    contracta,
    telNo,
    contractaName,
    newName,
newDepart,
newTarea,
newForm,
estado,
urgencia,
creator,

images,
    } = paraForm;



  const auth = getAuth();
  const navigate = useNavigate();

 function onChange(e) {
    let boolean = null;
    if (e.target.value === "true") {
      boolean = true;
    }
    if (e.target.value === "false") {
      boolean = false;
    }
    // Files
    if (e.target.files) {
      SetParaForm((prevState) => ({
        ...prevState,
        images: e.target.files,
      }));
    }
    // Text/Boolean/Number
    if (!e.target.files) {
      SetParaForm((prevState) => ({
        ...prevState,
        [e.target.id]: boolean ?? e.target.value,
      }));
    }
  }




  /// crear funcion que generar numero de Orden de trabajo para cada trabajo creado.
  function generate(n) {
    var add = 1,
      max = 12 - add;

    if (n > max) {
      return generate(max) + generate(n - max);
    }

    max = Math.pow(10, n + add);
    var min = max / 10;
    var number = Math.floor(Math.random() * (max - min + 1)) + min;

    return ("" + number).substring(add);
  }



  const createJob = async () => {
   
       setLoading(true);

    if (images.length > 10) {
      setLoading(false);
      toast.error("No se permite mas de 10 imagenes");
      return;
    }
    

    async function storeImage(image) {
      return new Promise((resolve, reject) => {
        const storage = getStorage();
        const filename = `${auth.currentUser.uid}-${image.name}-${uuidv4()}`;
        const storageRef = ref(storage, filename);
        const uploadTask = uploadBytesResumable(storageRef, image);
        uploadTask.on(
          "state_changed",
          (snapshot) => {
            const progress =
              (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
            switch (snapshot.state) {
              case "paused":
                break;
              case "running":
                break;
                default:console.log("Something is wrong");
            }
          },
          (error) => {
            reject(error);
          },
          () => {
            getDownloadURL(uploadTask.snapshot.ref).then((downloadURL) => {
              resolve(downloadURL);
            });
          }
        );
      });
    }

    const imgUrls = await Promise.all(
      [...images].map((image) => storeImage(image))
    ).catch((error) => {
      setLoading(false);
      toast.error("Imagenes subido");
      return;
    });

  


    await addDoc(workCollectionRef, {
      name: newName,
      NameDepart: newDepart,
      Tareas: newTarea,
      TipoTrabajo: newForm,
      timestamp: serverTimestamp(),
      NumeroOT: generate(5),
      EstadoTrabajo: estado,
      Emergencia: urgencia,
      address: address,
      TelNo: telNo,
      Images:imgUrls,
      Creator: creator,
      contractaName:contractaName,
      timeStart: startDate.toString(),
      timeFinish: endDate.toString(),
      userRef: auth.currentUser.uid,
    })
      .then(() => {
        alert("Se ha creado orden de trabajo correctamente👍");
       

      })
      .catch((error) => {
        alert(error.message);
      });
  
  };


  const handleSubmit = (e) => {
    e.preventDefault();

    createJob();
     navigate("/WorkList");
  };



  return (
    <div className="App">
      <Nav className="justify-content-center" activeKey="/Home">
        <Nav.Item>
          <Nav.Link
            href="/WorkList"
            className="bg-indigo-500 shadow-lg shadow-indigo-500/50 px-8 m-3 border-gray-300 rounded transition duration-150 ease-in-out"
          >
            Listas de Trabajos
          </Nav.Link>
        </Nav.Item>
        <Nav.Item>
          <Nav.Link
            href="/MyForms"
            className="bg-indigo-500 shadow-lg shadow-indigo-500/50 px-8 m-3 border-gray-300 rounded transition duration-150 ease-in-out"
          >
            Contactar con Nosotros
          </Nav.Link>
        </Nav.Item>
        <Nav.Item>
          <Nav.Link
            href="/MyWorkList"
            className="bg-indigo-500 shadow-lg shadow-indigo-500/50 px-8 m-3 border-gray-300 rounded transition duration-150 ease-in-out"
          >
            Mis listas de Trabajo Personal
          </Nav.Link>
        </Nav.Item>
      </Nav>

      <div className="bg-slate-400 bg-clip-border p-6 border-4 border-violet-300 border-dashed">
        <Form onSubmit={handleSubmit}>
          <InputGroup className="mb-3 text-center ps-5 pe-5 ">
            <InputGroup.Text>Nombre</InputGroup.Text>
            <Form.Control
              aria-label="Entra tu Nombre"
              id="newName"
               value={newName}
              onChange={onChange}
            />
          </InputGroup>
                <InputGroup className="mb-3 text-center ps-5 pe-5 ">
            <InputGroup.Text>Creator</InputGroup.Text>
            <Form.Control
              aria-label="Entra Nombre de creador de Orden"
               id="creator"
                value={creator}
               onChange={onChange}
            />
          </InputGroup>

          <InputGroup className="mb-3 text-center ps-5 pe-5 ">
            <InputGroup.Text>Nombre de Departamento</InputGroup.Text>
            <Form.Control
              aria-label="Entra Nombre de Departamento aqui"
              id="newDepart"
               value={newDepart}
               onChange={onChange}
            />
          </InputGroup>

          <Form.Group className="ps-5 pe-5" controlId="formGridState">
            <Form.Label>Seleciona tipo de trabajo contigo</Form.Label>
            <Form.Select
              defaultValue="..."
               onChange={onChange}
               id="newForm"
                value={newForm}
            >
              <option>....</option>
              <option>Preventivo</option>
              <option>Corectivo</option>
              <option>Predictivo</option>
              <option>Mejoras</option>
              <option>Propuesta</option>
            </Form.Select>
          </Form.Group>
          <div className="tareas">
            <InputGroup className="mt-5 mb-4 pe-5 ps-5">
              <InputGroup.Text>Discripcion de trabajo</InputGroup.Text>
              <Form.Control
                as="textarea"
                aria-label="Escribir Discripcion de trabajo"
                 id="newTarea"
                  value={newTarea}
                 onChange={onChange}
              />
            </InputGroup>
          </div>
          <div>
            <Form.Group className="ps-5 pe-5" controlId="formGridState">
              <Form.Label>Estado de Trabajo</Form.Label>
              <Form.Select
                defaultValue="..."
                 id="estado"
                  value={estado}
                onChange={onChange}
              >
                <option>....</option>
                <option>Abierta</option>
                <option>En curso</option>
                <option>Cerrado</option>
                <option>Planificado</option>
              </Form.Select>
            </Form.Group>
          </div>

          <div>
            <Form.Group className="ps-5 pe-5" controlId="formGridState">
              <Form.Label>Urgencia</Form.Label>
              <Form.Select
                defaultValue="..."
                 id="urgencia"
                 onChange={onChange}
              >
                <option>....</option>
                <option>Normal</option>
                <option>Planificado</option>
                <option>Urgente</option>
              </Form.Select>

              <p className="text-center font-semibold mt-3"> Contrata </p>
              <div className="flex mb-6">
                <button
                  type="button"
                  id="contracta"
                  value={true}
                  onClick={onChange}
                  className={`mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full mb-3 ${
                    !contracta
                      ? "bg-white text-black"
                      : "bg-slate-600 text-white"
                  }`}
                >
                  Si
                </button>
                <button
                  type="button"
                  id="contracta"
                  value={false}
                  onClick={onChange}
                  className={`ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full mb-3 ${
                    contracta
                      ? "bg-white text-black"
                      : "bg-slate-600 text-white"
                  }`}
                >
                  No
                </button>
              </div>
              {contracta && (
                <div>
                  <div>
                    <p className="text-center font-semibold mt-3">
                     Nombre de Contrata
                    </p>
                    <input
                      type="text"
                      id="contractaName"
                      value={contractaName}
                      onChange={onChange}
                      required
                      className=" m-3 w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
                    />
                  </div>
                  <div>
                    <p className="text-center font-semibold mt-3"> ADDRESS </p>
                    <input
                      type="text"
                      id="address"
                      value={address}
                      onChange={onChange}
                      required
                      className=" m-3 w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
                    />
                  </div>
                  <div>
                    <p className="text-center font-semibold mt-3">
                      {" "}
                      Numero de Telefono{" "}
                    </p>
                    <input
                      type="text"
                      id="telNo"
                      value={telNo}
                      onChange={onChange}
                      required
                      className=" m-3 w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
                    />
                  </div>
                </div>
              )}
            </Form.Group>
          </div>

          <div className="mb-3 mt-3  ps-5 pe-5 ">
            Elegir Fecha de Inicio:
            <DatePicker
              selected={startDate}
              selectsStart
              startDate={startDate}
              endDate={endDate}
              onChange={(date) => setStartDate(date)}
            />
          </div>

          <div className="mb-3 mt-3  ps-5 pe-5 ">
            Elegir Fecha para Terminar:
            <DatePicker
              selected={endDate}
              selectsEnd
              startDate={startDate}
              endDate={endDate}
              minDate={startDate}
              onChange={(date) => setEndDate(date)}
            />
          </div>

          
        <div className="mb-6">
          <p className="text-lg font-semibold">Images</p>
          <p className="text-gray-600">
            Primer imagen sera la cubierta (max 10)
          </p>
          <input
            type="file"
            id="images"
            onChange={onChange}
            accept=".jpg,.png,.jpeg"
            multiple
            required
            className="w-full px-3 py-1.5 text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:bg-white focus:border-slate-600"
          />
        </div>
          

          <div className="text-center">
            <Button
              className="mb-5 mt-3 ps-5 mx-auto"
              variant="primary"
              type="submit"
            >
              Crear Orden
            </Button>
          </div>
        </Form>
      </div>
    </div>
  );
}


