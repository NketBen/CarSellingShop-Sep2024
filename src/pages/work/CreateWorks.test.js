

import React from "react";
import {render, screen} from "@testing-library/react";
import '@testing-library/jest-dom'
import { BrowserRouter } from "react-router-dom";
import renderer from 'react-test-renderer'

//import CreateWorks from './MyForms';
import CreateWorks from './CreateWorks';

//hacemos test para comprobar que el elemento h1  con texto "Formulario De Contacto" existe 

it('render a message', ()=> {
    render(<BrowserRouter><CreateWorks message="Images"/></BrowserRouter>);
    const messageElement=screen.getByText(/Images/i);
    expect(messageElement).toBeInTheDocument();

});

//hacemos test para comprobar que el componente  MyForm.js no tiene comportamiento inesperada.

it('render correctly',()=>{
    const component=renderer.create(<BrowserRouter><CreateWorks/></BrowserRouter>);
   const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});