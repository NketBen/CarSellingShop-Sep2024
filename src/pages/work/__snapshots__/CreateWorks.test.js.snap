// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div
  className="App"
>
  <div
    className="bg-slate-400 bg-clip-border p-6 border-4 border-violet-300 border-dashed"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="mb-3 text-center ps-5 pe-5  input-group"
      >
        <span
          className="input-group-text"
        >
          Nombre
        </span>
        <input
          aria-label="Entra tu Nombre"
          className="form-control"
          id="newName"
          onChange={[Function]}
          value=""
        />
      </div>
      <div
        className="mb-3 text-center ps-5 pe-5  input-group"
      >
        <span
          className="input-group-text"
        >
          Creator
        </span>
        <input
          aria-label="Entra Nombre de creador de Orden"
          className="form-control"
          id="creator"
          onChange={[Function]}
          value=""
        />
      </div>
      <div
        className="mb-3 text-center ps-5 pe-5  input-group"
      >
        <span
          className="input-group-text"
        >
          Nombre de Departamento
        </span>
        <input
          aria-label="Entra Nombre de Departamento aqui"
          className="form-control"
          id="newDepart"
          onChange={[Function]}
          value=""
        />
      </div>
      <div
        className="ps-5 pe-5"
      >
        <label
          className="form-label"
          htmlFor="formGridState"
        >
          Seleciona tipo de trabajo contigo
        </label>
        <select
          className="form-select"
          defaultValue="..."
          id="newForm"
          onChange={[Function]}
          value=""
        >
          <option>
            ....
          </option>
          <option>
            Preventivo
          </option>
          <option>
            Corectivo
          </option>
          <option>
            Predictivo
          </option>
          <option>
            Mejoras
          </option>
          <option>
            Propuesta
          </option>
        </select>
      </div>
      <div
        className="tareas"
      >
        <div
          className="mt-5 mb-4 pe-5 ps-5 input-group"
        >
          <span
            className="input-group-text"
          >
            Discripcion de trabajo
          </span>
          <textarea
            aria-label="Escribir Discripcion de trabajo"
            className="form-control"
            id="newTarea"
            onChange={[Function]}
            value=""
          />
        </div>
      </div>
      <div>
        <div
          className="ps-5 pe-5"
        >
          <label
            className="form-label"
            htmlFor="formGridState"
          >
            Estado de Trabajo
          </label>
          <select
            className="form-select"
            defaultValue="..."
            id="estado"
            onChange={[Function]}
            value=""
          >
            <option>
              ....
            </option>
            <option>
              Abierta
            </option>
            <option>
              En curso
            </option>
            <option>
              Cerrado
            </option>
            <option>
              Planificado
            </option>
          </select>
        </div>
      </div>
      <div>
        <div
          className="ps-5 pe-5"
        >
          <label
            className="form-label"
            htmlFor="formGridState"
          >
            Urgencia
          </label>
          <select
            className="form-select"
            defaultValue="..."
            id="urgencia"
            onChange={[Function]}
          >
            <option>
              ....
            </option>
            <option>
              Normal
            </option>
            <option>
              Planificado
            </option>
            <option>
              Urgente
            </option>
          </select>
          <p
            className="text-center font-semibold mt-3"
          >
             Contrata 
          </p>
          <div
            className="flex mb-6"
          >
            <button
              className="mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full mb-3 bg-white text-black"
              id="contracta"
              onClick={[Function]}
              type="button"
              value={true}
            >
              Si
            </button>
            <button
              className="ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full mb-3 bg-slate-600 text-white"
              id="contracta"
              onClick={[Function]}
              type="button"
              value={false}
            >
              No
            </button>
          </div>
        </div>
      </div>
      <div
        className="mb-3 mt-3  ps-5 pe-5 "
      >
        Elegir Fecha de Inicio:
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className=""
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={false}
              type="text"
              value="08/27/2024"
            />
          </div>
        </div>
      </div>
      <div
        className="mb-3 mt-3  ps-5 pe-5 "
      >
        Elegir Fecha para Terminar:
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              className=""
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={false}
              type="text"
              value="08/27/2024"
            />
          </div>
        </div>
      </div>
      <div
        className="mb-6"
      >
        <p
          className="text-lg font-semibold"
        >
          Images
        </p>
        <p
          className="text-gray-600"
        >
          Primer imagen sera la cubierta (max 10)
        </p>
        <input
          accept=".jpg,.png,.jpeg"
          className="w-full px-3 py-1.5 text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:bg-white focus:border-slate-600"
          id="images"
          multiple={true}
          onChange={[Function]}
          required={true}
          type="file"
        />
      </div>
      <div
        className="text-center"
      >
        <button
          className="mb-5 mt-3 ps-5 mx-auto btn btn-primary"
          disabled={false}
          type="submit"
        >
          Crear Orden
        </button>
      </div>
    </form>
  </div>
</div>
`;
